rfcs/20220608-make-opdesc-and-md-opaque/README.md

```diff
@@ -0,0 +1,1186 @@
+        const_dnnl_op_desc_t op_desc, const_dnnl_primitive_attr_t attr,
+        dnnl_engine_t engine, const_dnnl_primitive_desc_t hint_forward_primitive_desc);
+```
+
+```cpp
+// Primitive descriptor iterator.
+dnnl_status_t dnnl_primitive_desc_iterator_create(
+        dnnl_primitive_desc_iterator_t *iterator,
+        const_dnnl_op_desc_t op_desc, const_dnnl_primitive_attr_t attr,
+        dnnl_engine_t engine, const_dnnl_primitive_desc_t hint_forward_primitive_desc);
+```
+
+This means that removing operation descriptor will require adjust both APIs, which
+will lead to duplication. To avoid the duplication the proposal is to add iterator
+capabilities to primitive descriptor and remove primitive descriptor iterator from API.
+
+From API perspective, there will be introduced a new API for primitive descriptor:
+```cpp
+dnnl_status_t dnnl_primitive_desc_next_impl(dnnl_primitive_desc_t primitive_desc);
+```
+This will be aligned with C++ API, where primitive descriptor provides `next_impl()`
+API.
+
+From implementation perspective, `primitive_desc_iface_t` will be adjusted so that
+it contains primitive descriptor iterator instead of primitive descriptor itself.
+
+The current API for creating primitive descriptor will be replaced with a set of the
+following APIs:
+```cpp
+// Depending on primitive kind and propagation kind the API will take different
+// arguments.
+dnnl_<primitive kind>_[propagation kind]_primitive_desc_create(...);
+```
+
+#### C++ API
+
+Concat, sum and reorder primitives don't have operation descriptors, all parameters
+are passed directly to primitive descriptor constructor.
+All other primitives will be aligned with concat, sum and reorder.
+
+* All nested `desc` classes will be removed
+* Each primitive descriptor class will get constructors that take the arguments that
+were previously passed to the `desc` classes constructors
+
+Iterator API will stay unchanged.
+
+### Querying Operation Descriptor Information
+
+Primitive descriptor will not be able to return an operation descriptor therefore
+the corresponding query will be removed.
+
+#### C API
+The proposal for query capabilities described in the option 1 will be used with the
+following changes:
+* The API for querying information from `op_desc_t` will not be added
+* The existing `dnnl_primitive_desc_query` API will handle all new queries
+
+#### C++ API
+The proposal for query capabilities described in the option 1 will be used with the
+following changes:
+1. No `query_helper` is needed, all the queries will be implemented in
+`primitive_desc_base`
+2. Query API for each primitive will be implemented in each primitive descriptor
+instead of operation descriptor
+
+## Operation Descriptor: Conclusion
+
+### Option 1
+* This option allows to hide implementation details of an operation descriptor
+and has little impact on users of the C++ API
+* On the other hand, there will be an opaque abstraction that doesn't bring any value
```

Comment:
There is some value in having `op_desc`. The good thing about `op_desc` is that user can catch issue related to invalid data types, algorithms, strides, etc. earlier. In case there is no `op_desc` and only `pd` all issues related to both mathematical operation configuration and unimplemented configuration of attributes and target engine should be detected in this API. It also requires user to create an engine before creating a `pd`.
Right now mostly `op_desc` returns `invalid_arguments`, while `pd` returns `unimplemented` if there is an issue.

Reply:
> The good thing about op_desc is that user can catch issue related to invalid data types, algorithms, strides, etc. earlier.

I agree, but why do you think this is useful?
Returning `invalid_arguments` upon `op_desc` creation indicates that there is a fundamental issue with using oneDNN API (e.g. source and destination memory have different number of dimensions, axis is negative, etc). In other words, this means that oneDNN API is used incorrectly and such issues should be address at integration time and are not supposed to be handled at runtime.
The `unimplemented ` status, on the contrary, is supposed to be handled at runtime.

This behavior will be preserved after removing operation descriptor.